geo/src/algorithm/interior_point.rs

```diff
@@ -0,0 +1,787 @@
+                        .map(|pt| (pt, pt.euclidean_distance(&centroid)))
+                })
+                .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Less))
+                .map(|(pt, _distance)| pt)
+        } else {
+            None
+        }
+    }
+}
+
+fn polygon_interior_point_with_segment_length<T: GeoFloat>(
+    polygon: &Polygon<T>,
+) -> Option<(Point<T>, T)> {
+    // special-case a one-point polygon since this algorithm won't otherwise support it
+    if polygon.exterior().0.len() == 1 {
+        return Some((polygon.exterior().0[0].into(), T::zero()));
+    }
+
+    let two = T::one() + T::one();
+
+    let bounds = if let Some(maybe_bounds) = polygon.bounding_rect() {
+        maybe_bounds
+    } else {
+        return None;
+    };
+
+    // use the midpoint of the bounds to scan, unless that happens to match any vertices from
+    // polygon; if it does, perturb the line a bit by averaging with the Y coordinate of the
+    // next-closest-to-center vertex if possible, to reduce the likelihood of colinear
+    // intersections
+    let mut y_mid = (bounds.min().y + bounds.max().y) / two;
+    if polygon.coords_iter().any(|coord| coord.y == y_mid) {
+        let next_closest = polygon
+            .coords_iter()
+            .filter_map(|coord| {
+                if coord.y == y_mid {
+                    None
+                } else {
+                    Some((coord.y, (coord.y - y_mid).abs()))
+                }
+            })
+            .min_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Less));
+        if let Some((closest, _)) = next_closest {
+            y_mid = (y_mid + closest) / two
+        }
+    };
+
+    let scan_line = Line::new(
+        Coordinate {
+            x: bounds.min().x,
+            y: y_mid,
+        },
+        Coordinate {
+            x: bounds.max().x,
+            y: y_mid,
+        },
+    );
+
+    let lines = polygon
+        .lines_iter()
+        .chain(std::iter::once(scan_line.clone()));
+
+    let mut intersections: Vec<SweepPoint<T>> = Vec::new();
+    for (l1, l2, inter) in Intersections::from_iter(lines) {
+        if !(l1 == scan_line || l2 == scan_line) {
+            continue;
+        }
+        match inter {
+            LineIntersection::Collinear { intersection } => {
+                intersections.push(SweepPoint::from(intersection.start));
+                intersections.push(SweepPoint::from(intersection.end));
+            }
+            LineIntersection::SinglePoint { intersection, .. } => {
+                intersections.push(SweepPoint::from(intersection));
+            }
+        }
+    }
+    intersections.sort();
+
+    let mut segments = Vec::new();
+    let mut intersections_iter = intersections.iter().peekable();
+    while let (Some(start), Some(end)) = (intersections_iter.next(), intersections_iter.peek()) {
+        let length = end.x - start.x;
+        let midpoint = Point::new((start.x + end.x) / two, start.y);
```

Comment:
I added a better explanation of the algorithmic approach, with a rough sketch, to the PR body above ^. But the gist of it: we're intersecting a horizontal line with the edge lines of the polygon, and then iterating over all the points where that horizontal line and the edge lines intersect (we're calculating the full intersection set, but then throwing out all the intersection points where neither of the source geometries is our horizontal split line in the preceding loop). As such, all of these points should fall along the same horizontal line, so they should all have the same Y coordinate. To maybe make this clearer, I changed the spelling in this bit of the function to use the previously-calculated Y value instead of `start.y`; this should be functionally identical, but maybe makes the intent clearer?